File: packages/dotnet/src/Internal/Wire.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace Opensms.Internal
{
    /// <summary>
    /// Helpers that turn SDK shapes into exact paths and query strings. Not part
    /// of the public API.
    /// </summary>
    internal static class Wire
    {
        /// <summary>URL-escape one path segment after checking it is non-empty (argument error, no request).</summary>
        public static string Seg(string? value, string name)
        {
            if (string.IsNullOrWhiteSpace(value))
                throw new ArgumentException($"{name} is required", name);
            return Uri.EscapeDataString(value);
        }

        /// <summary>Format an instant as RFC 3339 UTC (<c>2026-09-24T10:00:00Z</c>).</summary>
        public static string FormatTimestamp(DateTimeOffset value)
            => value.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);

        /// <summary>Accumulates query parameters, skipping unset values.</summary>
        public sealed class QueryBuilder
        {
            private readonly List<KeyValuePair<string, string>> _parts = new List<KeyValuePair<string, string>>();

            public QueryBuilder Add(string key, string? value)
            {
                if (value != null) _parts.Add(new KeyValuePair<string, string>(key, value));
                return this;
            }

            public QueryBuilder Add(string key, int? value)
                => Add(key, value?.ToString(CultureInfo.InvariantCulture));

            public QueryBuilder Add(string key, long? value)
                => Add(key, value?.ToString(CultureInfo.InvariantCulture));

            public QueryBuilder Add(string key, bool? value)
                => Add(key, value == null ? null : (value.Value ? "true" : "false"));

            public QueryBuilder Add(string key, DateTimeInput? value)
                => Add(key, value?.Value);

            /// <summary>List values are joined with commas (<c>countries=KE,NG</c>).</summary>
            public QueryBuilder Add(string key, IEnumerable<string>? values)
                => Add(key, values == null ? null : string.Join(",", values));

            /// <summary>Returns <c>""</c> when empty, else <c>?a=1&amp;b=2</c> with keys and values escaped.</summary>
            public override string ToString()
            {
                if (_parts.Count == 0) return "";
                var sb = new StringBuilder("?");
                for (var i = 0; i < _parts.Count; i++)
                {
                    if (i > 0) sb.Append('&');
                    sb.Append(Uri.EscapeDataString(_parts[i].Key)).Append('=').Append(Uri.EscapeDataString(_parts[i].Value));
                }
                return sb.ToString();
            }
        }

        public static QueryBuilder Query() => new QueryBuilder();

        /// <summary>Start a query with the common <c>limit</c> and <c>cursor</c> keys.</summary>
        public static QueryBuilder Query(ListParams? p) => new QueryBuilder().Add("limit", p?.Limit).Add("cursor", p?.Cursor);
    }
}

File: packages/java/src/main/java/io/opensms/Wire.java
package io.opensms;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;

/** Internal helpers for building JSON request bodies. Not part of the public API. */
final class Wire {

    private Wire() {
    }

    /** A new ordered map for accumulating wire fields. */
    static Map<String, Object> map() {
        return new LinkedHashMap<>();
    }

    /** Put {@code value} under {@code key} only when non-null, so unset optionals are omitted. */
    static void put(Map<String, Object> m, String key, Object value) {
        if (value != null) {
            m.put(key, value);
        }
    }

    /** Fail fast with an argument error when a required value is missing. */
    static <T> T require(T value, String name) {
        if (value == null || (value instanceof String s && s.isEmpty())) {
            throw new IllegalArgumentException(name + " is required");
        }
        return value;
    }

    /** Validate a path id (non-empty, else an argument error before any request) and URL-escape it. */
    static String id(String value, String name) {
        if (value == null || value.isEmpty()) {
            throw new IllegalArgumentException(name + " is required");
        }
        return Query.seg(value);
    }

    /** RFC 3339 in UTC, for example {@code 2026-09-24T10:00:00Z}. */
    static String rfc3339(Instant t) {
        return t == null ? null : DateTimeFormatter.ISO_INSTANT.format(t);
    }

    /** RFC 3339 in UTC. */
    static String rfc3339(OffsetDateTime t) {
        return t == null ? null : rfc3339(t.withOffsetSameInstant(ZoneOffset.UTC).toInstant());
    }
}
